landoapi/transplants.py

```diff
@@ -322,8 +341,15 @@ def check_landing_blockers(
     stack_data,
     landable_paths,
     landable_repos,
+    reviewers,
+    users,
+    projects,
     *,
-    user_blocks=[user_block_no_auth0_email, user_block_scm_level]
+    user_blocks=[
+        user_block_no_auth0_email,
+        user_block_scm_level,
+        repo_block_approval_required,
```

Comment:
This isn't quite the right place for this check, but looking at things again, the subtly of where blockers should go isn't well documented or obvious. Here's a bit of a high level description of the blocker/warning flow, as well as hopefully some context for why it's that way:

## Background
There are three stages of checking that happen:
1. **Pre-landing public** checks.
2. **Pre-landing dryrun** checks.
3. **On landing request** checks.

**Pre-landing public** checks are the checks we run whenever a stack is visited. The idea here is that these checks are purely about the state of Phabricator and indicate if a revision / series is allowed to land. They should not depend on *who* is making the request to land, or things like in progress landings, only the state of the revision / series / stack in Phabricator. When viewing a stack page in Lando, this is what decides if a revision / series is blocked, and what you may select to land. These are purely blocker checks and are implemented by the `calculate_landable_subgraphs()` function.

`calculate_landable_subgraphs()` does a bunch of checking about the graphs structure, making sure you can't request a landing on something with open parents for example. What matters most here though is the `other_checks` argument it takes, which is described in the docstring:
> An iterable of callables which will be executed for each revision to determine if it should be blocked. These checks must not rely on the structure the stack graph takes, instead they may check only properties specific to a single revision.

This is where blockers based on revision data may be passed in order to restrict landing in certain states. Current checks are:
- Has the author indicated they're planning changes to this revision? (`check_author_planned_changes`)
- Does the uploaded diff contain the commit author information? (`check_diff_author_is_known`)

**Pre-landing dryrun** and **On landing request** checks are the same checks, just run at different times. The dryrun happens to generate the landing preview and make sure the warnings you saw before landing match the warnings when the landing is initiated. In the context of your change we can ignore the difference here and focus on the actual checks themselves.

So, the **On landing request** checks include both warnings and blockers. The blockers are implemented by the `check_landing_blockers()` function you're modifying here, and are either inherent to the function, or provided in the `user_blocks` list. Things inherent to the function include:
- Are the diffs we are attempting to land right obsolete?
- Is there already a landing queued for this stack?
- Is the requested landing a proper landable path (as generated by `calculate_landable_subgraphs()`).

As you can see, this is where state that lives outside of Phabricator comes into play (such as landings already being in the queue). These checks probably won't change very often, and they are more about "Can this specific request go through?" rather than "Is this specific set of revisions / diffs on Phabricator something that's okay to land?"

The checks provided through `user_blocks` are similar, but focus on the user making the landing. So while something might be okay to land, the specific user making this request might be blocked from landing it because:
- We couldn't get their email from auth0 (`user_block_no_auth0_email`)
- They don't have the required scm level to land here (`user_block_scm_level`).

These user checks are provided as arguments to make testing easier and because they will change in the future as we switch away from scm_level for permission to land (using something like module ownership / having review and being the revision author, etc)

The other piece of the **On landing request** checks are the warnings. These again look at Phabricator state but live here as they can be about the request to land, and not just specific revision details. Many of the warnings are also expensive to generate so calculating them for a full stack, rather than just the specific substack that's landing could be problematic.

## Where to go from here
We want to transition your `repo_block_approval_required` check to be **pre-landing public** check. It's based only on Phabricator state information, and should prevent a user from selecting the revision from landing. We'll want this to be part of the `calculate_landable_subgraphs()` checks.

You should write a function to be passed into `other_checks`, which will look for an approval from the appropriate group if the repo requires approval. These checks are already called with all the data you'll need to make this decision `check(revision=revision, diff=diff, repo=repo)`. Rather than searching the names of all the reviewers and comparing to the group, take a look at what `get_secure_project_phid()`, `get_checkin_project_phid()` etc. do and make something similar for the release-managers group. That way you can just grab the PHID and check it against the PHID already attached to the revision object.

Reply:
I just did that change.